### PR TITLE
Add Repl adapter

### PR DESCRIPTION
**Pull request by YeeeeeHan**
**NOTE**

> - If you would like to add a `volume` adapter please submit the PR [here](https://github.com/DefiLlama/adapters).
> - If you would like to add a `liquidations` adapter, please refer to [this readme document](https://github.com/DefiLlama/DefiLlama-Adapters/tree/main/liquidations) for details.

1. Once your adapter has been merged, it takes time to show on the UI. If more than 24 hours have passed, please let us know in Discord.
2. Please enable "Allow edits by maintainers" while putting up the PR.
3. Sorry, We no longer accept fetch adapter for new projects, we prefer the tvl to computed from blockchain data, if you have trouble with creating a the adapter, please hop onto our discord, we are happy to assist you.
4. Please fill the form below  **only if the PR is for listing a new protocol** else it can be ignored/replaced with reason/details about the PR
5. **For updating listing info** It is a different repo, you can find your listing in this file: https://github.com/DefiLlama/defillama-server/blob/master/defi/src/protocols/data2.ts, you can  edit it there and put up a PR
6. Do not edit/push `package-lock.json` file as part of your changes, we use lockfileVersion 2, and most use v1 and using that messes up our CI
7. No need to go to our discord and announce that you've created a PR, we monitor all PRs and will review it asap

---
##### Name (to be shown on DefiLlama):
Repl

##### Twitter Link:
-

##### List of audit links if any:
- [Zokyo](https://docs.repl.fi/assets/files/Repl_Zokyo_audit_report_2023-75580faff2dc76b2dd8d98cde41cf2dc.pdf) 
- [SlowMist](https://docs.repl.fi/assets/files/SlowMist_Audit_Report-47a5f3b63df90c99a4aa4d39ce367307.pdf)

##### Website Link:
https://repl.fi/

##### Logo (High resolution, will be shown with rounded borders):
repl_logo

##### Current TVL:
$3,839,225

##### Treasury Addresses (if the protocol has treasury)
[0xD51Cb0FA9A91f156a80188a18f039140704b8df7](https://filfox.info/en/address/0xD51Cb0FA9A91f156a80188a18f039140704b8df7)

##### Chain:
Filecoin - 314

##### Short Description (to be shown on DefiLlama):
Repl is the first repledging Protocol for pledged FIL assets. It aims to enhance the efficiency of pledged FIL, remove idiosyncratic risks among Filecoin storage providers, and lower the complexity for people to earn Filecoin mining return.

##### Token address and ticker if any:
Token Address: [0xAaa93ac72bECfbBc9149f293466bbdAa4b5Ef68C](https://filfox.info/en/address/0xAaa93ac72bECfbBc9149f293466bbdAa4b5Ef68C)
Ticker: pFIL

##### Category (full list at https://defillama.com/categories) *Please choose only one:
[Infrastructure](https://defillama.com/protocols/Infrastructure)

##### Oracle used (Chainlink/Band/API3/TWAP or any other that you are using):
-

##### forkedFrom (Does your project originate from another project):
-

##### methodology (what is being counted as tvl, how is tvl being calculated):
TVL = value of eligible assets of all nodes + outstanding pFIL value + value of non-pFIL liquidity in all participating DEXs, denominated in USD

##### Github org/user (Optional, if your code is open source, we can track activity):
-

### CODE REVIEW

File: projects/repl/index.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        totalPFILSupply,
+        wpFILprice,
+        Number(ethers.utils.formatEther(PFILperToken)),
+        FILprice
+      );
+      const tvl = totalEligibleAssetsValue
+        .add(poolWFILBal)
+        .add(totalSupplyValue);
+
+      // console.log(
```
Comment (Viyozc) [inline projects/repl/index.js:133]: remove console?

Comment (YeeeeeHan) [reply projects/repl/index.js:133]: Ok

**Comment by llamatester** [comment]
The adapter at projects/repl exports TVL: 
        
 
 
    filecoin                  3.88 M
    
    total                    3.88 M

**Comment by waynebruce0x** [comment]
Can you explain what totalEligibleAssets? How is it different to total supply of pFIL?

**Comment by YeeeeeHan** [comment]
> Can you explain what totalEligibleAssets? How is it different to total supply of pFIL?

@waynebruce0x When a miner in filecoin participates in our protocol, they will transfer ownership of all the assets in the miner to the protocol. Eligible Assets refer to the assets within the miner, and thus total Eligible Assets refer to all of the eligible assets of all miners participating in the protocol.

While pFIL supply is the total value of PFIL in circulation. These pFIL tokens are minted as result of miners pledging their assets to the protocol.

**Comment by g1nt0ki** [comment]
@YeeeeeHan thanks for the PR, I have refactored and merged your code here: https://github.com/DefiLlama/DefiLlama-Adapters/pull/8600/files

I dont think we should count the pFIL supply as tvl as the tokens backing it are already counted as part of tvl

**Comment by YeeeeeHan** [comment]
GM @g1nt0ki , I would like to revive this discussion on the TVL calculation:
- When miners pledge their assets to the protocol, pFIL is not minted immediately, instead the pFIL allowance is calculated
At this point, the TVL increases by the `totalEligibleAsset` of the miner
- When pFIL is minted and sold to the market, the price of pFIL represents the value that the market locks in the Repl protocol in exchange for future cash flows of the miner.
- We can consider the value of pFIL as the external liquidity provided by pFIL holders to the storage providers. After all, storage providers have no incentive to mint pFIL if no one is willing to fund them.
- Looking at the DEX pools, we can consider all the locked pFIL and FIL as part of the protocol's TVL. As some pFIL are exchanged OTC, we cannot track all markets, but we do know that all pFIL in circulation is funded by FIL or other cryptocurrencies.  Therefore, we use the price reflected in the DEX pool as a way to estimate the value of funding for all pFIL. To avoid double counting of pFIL in the DEX pool and in circulation, we calculate TVL by summing up:
     - `Non-pFIL assets in known liquidity pools`
     - `Total Circulating Supply of pFIL`
     - `totalEligibleAsset`

**Comment by g1nt0ki** [comment]
@YeeeeeHan if we can compare this to ETH locked in Maker to mint DAI, we only count the ETH locked as tvl, DAI is not counted as tvl, even though we expect DAI value to be 1, it is market determined, the same argument what you made for pFIL?

**Pull request closed without merging**